Adds distance method from spherical LineString to Point

eikes opened the pull request:
This PR implements two new methods to the `SphericalLineStringMethods` module, both can only take a `SphericalPointImpl` as a parameter:

`#distance`
`#closest_point`

The implementation requires the closest_point method to also be present in the internal `RGeo::Geographic::SphericalMath::ArcXYZ` class. This in turn depends on the ability to project a point on the unit sphere onto the great arc of a different arc. This is implemented in the `project_point` method, it introduces the building block which the other methods depend on. The other methods are more or less trivial.

eikes committed:
Adds distance method from spherical LineString to Point

eikes committed:
Return s as closest_point when e equals s

eikes committed:
Apply style fixes to satisfy rubocop

keithdoggett reviewed (approved):
@eikes thank you for the PR. The tests are clear and the code looks good to me. Just one question. How hard would it be to implement linestring to linestring distance now? Seems like it should be an easy enough lift given that you've already done the hard part here. Not necessary to get this PR merged but if it's easy enough could make sense to do now too.

eis-ioki commented:
@keithdoggett I don't think the distance of two arcs on a sphere is ever anything other than the minimum of the distances of all enpoints to the other arc, unless they cross each other:

```ruby
def distance(arc_1, arc_2)
  return 0 if arc_1.crosses?(arc_2)
  [
    arc_1.s.distance(arc_2),
    arc_1.e.distance(arc_2),
    arc_2.s.distance(arc_1),
    arc_2.e.distance(arc_1)
  ].min
end
```

eikes commented:
@keithdoggett I don't think the distance of two arcs on a sphere is ever anything other than the minimum of the distances of all enpoints to the other arc, unless they cross each other:

```ruby
def distance(arc_1, arc_2)
  return 0 if arc_1.crosses?(arc_2)
  [
    arc_1.s.distance(arc_2),
    arc_1.e.distance(arc_2),
    arc_2.s.distance(arc_1),
    arc_2.e.distance(arc_1)
  ].min
end
```

eikes committed:
Create a real life test

eikes commented:
I am getting unexpected results here. I am reverting to draft status till I know why...

eikes committed:
Fix spherical arc containment for projected points near the great-circle plane

Fix test failure:

The actual bug was in the arc-membership check, where spherical_math.rb
required the projected point to satisfy plane membership with exact
equality. In the Seattle–Brussels–Reykjavik case, the projection is
correct and lands about 38.8 km away, but its residual against the
plane is about `1.1×10^−16`, so the code rejected it and fell back
to the Brussels endpoint, producing the 2129 km result.

eikes committed:
Fix spherical arc containment orientation checks

eikes committed:
Make spherical arc containment tolerance explicit

eikes commented:
@keithdoggett I have finally figured out why my real life test was failing. It now passes. 

The project_point method was return a point that, due to floating point math, was just a little bit off the plane. That in turn led to contains_point? returning false, which in the end meant that it didn't return a point on the arc, but rather an endpoint of the linestring.

eikes committed:
Update the great arc examples to use the closest point of reykjavik airport to the linestring from bern airport to seattle airport

BuonOmo commented:
> The project_point method was return a point that, due to floating point math, was just a little bit off the plane.

Interesting, it seems related to #395. We may have a global issue on handling floating point errors
